DM-34874: Refactor parquet formatter to use Arrow Tables natively, and add converters.

erykoff opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj commented on python/lsst/daf/butler/configs/datastores/formatters.yaml:
```diff
@@ -71,7 +71,9 @@ StampsBase: lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter
 Stamps: lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter
 BrightStarStamps: lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter
 AstropyTable: lsst.daf.butler.formatters.astropyTable.AstropyTableFormatter
-AstropyQTable: lsst.daf.butler.formatters.astropyTable.AstropyTableFormatter
```
Why is this removed?

erykoff replied:
I don't know.  Restoring it.

timj commented on python/lsst/daf/butler/delegates/arrowastropy.py:
```diff
@@ -0,0 +1,95 @@
+            return ArrowAstropySchema(composite)
+        elif componentName == "rowcount":
+            return len(composite)
+        else:
```
Doesn't need the else here since you want to always raise if you get to the end of the method.

timj commented on python/lsst/daf/butler/delegates/arrowastropy.py:
```diff
@@ -0,0 +1,95 @@
+        AttributeError
+            The component can not be found.
+        """
+        if componentName == "columns":
```
Can we use the new match/case here?

erykoff replied:
Sure, why not?

timj commented on python/lsst/daf/butler/delegates/arrownumpy.py:
```diff
@@ -0,0 +1,97 @@
+                f"Do not know how to retrieve component {componentName} from {get_full_type_name(composite)}"
+            )
+
+    def handleParameters(
```
There's an uncomfortable amount of code duplication between this method and the ones in the other delegates. Seems like only one line differs (the "if column not in") and the instance check (the python type for that could be a class attribute). I don't like that much repeated logic across multiple files.

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+
+    def write(self, inMemoryDataset: Any) -> None:
+        import numpy as np
+        import pandas as pd
```
Importing pandas just to check if we have been given a pandas dataframe is a tad annoying.

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+        import pandas as pd
+        from astropy.table import Table as astropyTable
+
+        location = self.makeUpdatedLocation(self.fileDescriptor.location)
```
move this line down to where it's needed (just before the `write_table`)

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
-            raw_columns = list(self._splitColumnnNames(len(indexes), self.file.metadata.schema.names))
-            self.columns = pd.MultiIndex.from_tuples(raw_columns, names=[f["name"] for f in indexes])
-        self.indexLevelNames = tuple(self.columns.names)
+            raise ValueError("Unsupported type of inMemoryDataset for ParquetFormatter.")
```
Maybe report the get_full_type_name in this message?

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+
+def arrow_to_pandas(arrow_table: pa.Table) -> pd.DataFrame:
+    """Convert a pyarrow table to a pandas DataFrame.
+    If the input table has ``pandas`` metadata in the schema it will be
```
Blank line before this one since this is a note and not the short summary. It should probably be attached to the parameter description below.

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+
+    Returns
+    -------
+    array : `numpy.ndarray` (N,)
```
Maybe add a short description of the shape of the numpy array here rather than using `N` where `N` has no context.

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+
+    Parameters
+    ----------
+    arrow_table : `pyarrow.Table`
```
Normally we'd have a short text of something like "The table to be converted." rather than no description at all (applies to all these parameters).

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+
+    @classmethod
+    def from_arrow(cls, schema: pa.Schema) -> DataFrameSchema:
+        """Convert an arrow schema into a DataFrameSchema.
```
backticks around `DataFrameSchema`?

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
-        ------
-        tuple : `tuple` of `str`
-            A multi-index column name tuple.
+        schema : `pyarrow.Schema`
```
Short description?

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+        return arrow_table.schema
+
+    def to_arrow_numpy_schema(self) -> ArrowNumpySchema:
+        """Convert to an ArrowNumpySchema.
```
backticks?

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+            return False
+
+        for name in self._schema.columns:
+            if not self._schema[name].unit == other._schema[name].unit:
```
Am I missing something subtle as to why this isn't using `!=`?

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, ArrowAstropySchema):
+            return False
```
`return NotImplemented` ?

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+
+    @classmethod
+    def from_arrow(cls, schema: pa.Schema) -> ArrowNumpySchema:
+        """Convert an arrow schema into an ArrowNumpySchema.
```
back ticks.

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+    Parameters
+    ----------
+    numpy_dtype : `numpy.dtype` or `list` [`str`]
+    """
```
parameter description text?

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
-    """Interface for reading and writing Pandas DataFrames to and from Parquet
-    files.
+    def to_arrow_astropy_schema(self) -> ArrowAstropySchema:
+        """Convert to an ArrowAstropySchema.
```
Backticks? (I'll stop now since there are more examples)

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+    else:
+        if not isinstance(columns, collections.abc.Mapping):
+            raise ValueError(
+                "Columns parameter for multi-index data frame must be a " "dictionary or list of tuples."
```
Say what type you got?

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+    Parameters
+    ----------
+    astropy_table : `astropy.table.Table`
+        Table to apply metdata
```
typo: metdata

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+    md_name = f"lsst::arrow::len::{name}"
+    strlen = default_length
+    metadata = schema.metadata if schema.metadata is not None else {}
+    if md_name.encode("UTF-8") in metadata:
```
can you say:

```python
if (encoded := md_name.encode("UTF-8")) in metadata:
    strlen = int(schema.metadata[encoded])
```
so you don't need to encode twice?

timj commented on python/lsst/daf/butler/formatters/parquet.py:
```diff
@@ -21,177 +21,842 @@
+    if dtype.type is np.str_:
+        metadata[f"lsst::arrow::len::{name}".encode("UTF-8")] = str(dtype.itemsize // 4)
+    elif dtype.type is np.bytes_:
+        metadata[f"lsst::arrow::len::{name}".encode("UTF-8")] = str(dtype.itemsize)
```
Is this called with a non-string dtype ever?

erykoff replied:
Yes, and then this is a no-op.  Is that a problem?

timj replied:
I wanted to know if the dangling else was okay. Maybe add a note to the docstring saying that all column types are handled but only string types are modified?

timj reviewed (approved):
Some minor cleanups noted and I worry about code duplication.

Thanks for the comprehensive tests with excellent code coverage.

erykoff committed:
Refactor parquet formatter to be based on arrow tables natively.

Add ArrowTable, ArrowAstropy, ArrowNumpy support, as well as
relevant schema types and converters.

erykoff committed:
Add delegates for ArrowTable, ArrowAstropy, and ArrowNumpy tables.

erykoff committed:
Update storageclass and formatter configs for ArrowTable etc.

erykoff committed:
Add tests for new parquet formatter.

erykoff committed:
Add changelog entry for arrow parquet support.
